django/core/paginator.py

```diff
@@ -236,3 +356,96 @@ def end_index(self):
         if self.number == self.paginator.num_pages:
             return self.paginator.count
         return self.number * self.paginator.per_page
+
+
+class AsyncPage:
+    def __init__(self, object_list, number, paginator):
+        self.object_list = object_list
+        self.number = number
+        self.paginator = paginator
+
+    def __repr__(self):
+        return "<Async Page %s>" % self.number
+
+    async def __aiter__(self):
+        if hasattr(self.object_list, "__aiter__"):
+            async for obj in self.object_list:
+                yield obj
+        else:
+            for obj in self.object_list:
+                yield obj
+
+    def __len__(self):
+        if not isinstance(self.object_list, list):
+            raise TypeError(
+                "AsyncPage.aget_object_list() must be awaited before calling len()."
+            )
+        return len(self.object_list)
+
+    def __reversed__(self):
+        if not isinstance(self.object_list, list):
+            raise TypeError(
+                "AsyncPage.aget_object_list() "
+                "must be awaited before calling reversed()."
+            )
+
+        return reversed(self.object_list)
+
+    def __getitem__(self, index):
+        if not isinstance(index, (int, slice)):
+            raise TypeError(
+                "AsyncPage indices must be integers or slices, not %s."
+                % type(index).__name__
+            )
+
+        if not isinstance(self.object_list, list):
+            raise TypeError(
+                "AsyncPage.aget_object_list() must be awaited before using indexing."
+            )
+        return self.object_list[index]
+
+    async def aget_object_list(self):
+        """
+        Returns self.object_list as a list.
+
+        This method must be awaited before AsyncPage can be
+        treated as a sequence of self.object_list.
+        """
+        if not isinstance(self.object_list, list):
```

Comment:
Can we add a test to cover when `aget_object_list` is called and `self.object_list` is already a list?
I can remove the if without any test failures